### DISCUSSION TITLE
CopyFileEx returns different results based on VisualStudio 2019 vs VisualStudio 2022

### DISCUSSION THREAD

**Discussion by RFBomb** [dotnet/csharplang#8046] [Q&A]
EDIT : Answer to this question here : https://github.com/dotnet/runtime/discussions/100840#discussioncomment-9070534
----------

I am not sure at all where to ask this question, so please let me know if there is a better forum for this. 

I am one of the maintainers of the RoboSharp library, and am working on a PR to bring some custom implementations into existence for consumers. One of those will effectively be a wrapper for CopyFileEx, or its planned to be if all goes well (which its not). 

The problem i have run into is that 2 of my unit tests fail when running via VS2019 Professional, and succeeds when running via VS2022 Community. `CopyFileEx_CopyFile` and `CopyFileEx_CopyFileAsync`  (ignore `CopyFileAsyncTest`)

![image](https://github.com/tjscience/RoboSharp/assets/20431767/02728551-3ec0-4fe7-8787-8b444958c238)

Test Overview:
- Fail the test if source file missing (succeeds)
- Write the source and destination file to disk (succeeds)
- Attempt to execute CopyFileEx using the 'FAIL_IF_EXISTS' value of 0x1 --- This is proving reliably inconsistent!
     -  VS 2022 ALWAYS throws an exception (as expected)
     - VS 2019 ALWAYS overwrites the destination ( This is incorrect behavior of CopyFileEx )

Relevant Code:

```
string sourceFile = GetRandomPath();
string destFile = GetRandomPath();

Console.WriteLine(string.Format("Source: {0}\nDestination: {1}", sourceFile, destFile));
if (File.Exists(sourceFile)) File.Delete(sourceFile);
Assert.ThrowsException<FileNotFoundException>(() => FileFunctions.CopyFile(sourceFile, destFile, CopyFileExOptions.FAIL_IF_EXISTS));

File.WriteAllText(sourceFile, "Test Contents");
File.WriteAllText(destFile, "Content to replace\n\nMoreContent");
Assert.IsTrue(File.Exists(sourceFile));
Assert.IsTrue(File.Exists(destFile));

// This is the test that is inconsistent
Assert.ThrowsException<IOException>(() => FileFunctions.CopyFile(sourceFile, destFile, CopyFileExOptions.FAIL_IF_EXISTS), "\nCopy Operation Succeeded when CopyFileExOptions.FAIL_IF_EXISTS was set"); 

```
Other relevant code :

```
[Flags]
public enum CopyFileExOptions : uint
{

    NONE = 0x0,

    FAIL_IF_EXISTS = 0x1,
}

        [DllImport("kernel32.dll", SetLastError = true, CharSet = CharSet.Ansi, ThrowOnUnmappableChar = true)]
        [return: MarshalAs(UnmanagedType.Bool)]
        private static extern bool CopyFileEx(
            string lpExistingFileName,
            string lpNewFileName,
            CopyProgressCallback lpProgressRoutine,
            object lpData,
            ref int pbCancel,
            uint dwCopyFlags);


public static bool CopyFile(
            string source,
            string destination,
            CopyFileExOptions flags = CopyFileExOptions.NONE,
            CopyProgressCallback progressCallback = null,
            IntPtr? data = null
            )
        {
            if (!File.Exists(source)) throw new FileNotFoundException("Source File Not Found.", source);
            new FileInfo(destination).Directory.Create();
            int cancel = 0;
            bool returnVal = CopyFileEx(source, destination, progressCallback, data, ref cancel, (uint)flags);
            if (returnVal is false) Win32Error.ThrowLastError(source, destination); // Method to get last error and throw
            return returnVal;
        }
```

PR I discovered the issue in : https://github.com/tjscience/RoboSharp/pull/193
Other files with relevant code : 
- Unit Tests --  https://github.com/RFBomb/RoboSharp/blob/AsyncCopy/RoboSharp.Extensions.UnitTests/CopyFileTests.cs
- DllImport --  https://github.com/RFBomb/RoboSharp/blob/AsyncCopy/RoboSharp.Extensions/CopyFileEx/FileFunctions.CopyFileAsync.cs
- Enum -- https://github.com/RFBomb/RoboSharp/blob/AsyncCopy/RoboSharp.Extensions/CopyFileEx/CopyFileExOptions.cs
- https://github.com/RFBomb/RoboSharp/blob/AsyncCopy/RoboSharp.Extensions/CopyFileEx/FileFunctions.Callbacks.cs
- https://github.com/RFBomb/RoboSharp/blob/AsyncCopy/RoboSharp.Extensions/CopyFileEx/Win32Error.cs

**Comment by CyrusNajmabadi** [comment]
Closing out.  You could consider asking dotnet/runtime about this.  This is the repo to discuss the C# language itself, and its evolution.  I would move your issue, but github seems to be buggy and i don't have the option to do it myself :(

**Reply by RFBomb** [reply]
Thanks for the response, I'll move this over there
